### PR TITLE
Pluggable SCM - get, create, update (#1508)

### PR DESCRIPTION
**Pull request by varshavaradarajan**
* Changed Configuration Property Builder to accomodate for all types of plugins.

### CODE REVIEW

File: server/webapp/WEB-INF/rails.new/app/controllers/api_v1/admin/pluggable_scms_controller.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      def update
+        result = HttpLocalizedOperationResult.new
+        @scm = ApiV1::Scms::PluggableScmRepresenter.new(SCM.new).from_hash(params)
+        pluggable_scm_service.updatePluggableScmMaterial(current_user, @scm, result)
+
+        json = ApiV1::Scms::PluggableScmRepresenter.new(@scm).to_hash(url_builder: self)
+        result.isSuccessful ? (render DEFAULT_FORMAT => json) : (render_http_operation_result(result, {data: json}))
+      end
+
+      def check_for_stale_request
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/controllers/api_v1/admin/pluggable_scms_controller.rb]: please mark these methods as `private` as a matter of hygine, it only takes a bad route to be able to reach this method.

File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/scms/pluggable_scm_representer.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      end
+
+      link :doc do
+        'http://api.go.cd/#pluggable-scms'
+      end
+      property :errors, exec_context: :decorator, decorator: ApiV1::Config::ErrorRepresenter, skip_parse: true, skip_render: lambda { |object, options| object.empty? }
+      property :id
+      property :name
+      property :auto_update
+      property :plugin_configuration,
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/presenters/api_v1/scms/pluggable_scm_representer.rb]: I know I'm deviating, but should we call this `plugin_metadata`, `configuration` and `plugin_configuration` are quite ambiguous.

**Comment by ketan** [comment]
I know that `PluginConfigurationPropertyRepresenter` was not part of this PR, but what happens when you submit both `value` and `encrypted` value to it?

**Comment by varshavaradarajan** [comment]
For ```PluginConfigurationPropertyRepresenter```, a configuration property object will be created with both the value and encrypted_value set. But this is not the property object that will be added to the configuration be it scm or pluggable task but rather the property object created out of this deserialized property object and the configuration preference or metadata which provides the options for each of the properties

File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/scms/pluggable_scm_representer.rb (hunk 2 above)
Comment (varshavaradarajan) [inline server/webapp/WEB-INF/rails.new/app/presenters/api_v1/scms/pluggable_scm_representer.rb]: makes sense. But in the model it is known as PluginConfiguration and Configuration anyway.

File: server/src/com/thoughtworks/go/config/update/CreateSCMConfigCommand.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    public void update(CruiseConfig modifiedConfig) throws Exception {
+        SCMs scms = modifiedConfig.getSCMs();
+        scms.add(globalScmConfig);
+        modifiedConfig.setSCMs(scms);
+    }
+
+    @Override
+    public boolean isValid(CruiseConfig preprocessedConfig) {
```
Comment (ketan) [inline server/src/com/thoughtworks/go/config/update/CreateSCMConfigCommand.java]: Is valid is duplicated between create and update commands, is there a way to remove that duplication?

File: server/src/com/thoughtworks/go/config/update/CreateSCMConfigCommand.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        return false;
+    }
+
+    @Override
+    public SCM getPreprocessedEntityConfig() {
+        return preprocessedGlobalScmConfig;
+    }
+
+    @Override
+    public boolean canContinue(CruiseConfig cruiseConfig) {
```
Comment (ketan) [inline server/src/com/thoughtworks/go/config/update/CreateSCMConfigCommand.java]: This too seems to be duplicated

File: server/src/com/thoughtworks/go/config/update/CreateSCMConfigCommand.java [hunk 5]
```diff
@@ -56,30 +41,16 @@ public void update(CruiseConfig modifiedConfig) throws Exception {
...
     }
 
     @Override
     public boolean canContinue(CruiseConfig cruiseConfig) {
-        if (!(goConfigService.isUserAdmin(currentUser)) || goConfigService.isGroupAdministrator(currentUser.getUsername())) {
-            result.unauthorized(LocalizedMessage.string("UNAUTHORIZED_TO_EDIT"), HealthStateType.unauthorised());
-            return false;
-        }
-        return true;
+        return super.canContinue(cruiseConfig);
```
Comment (ketan) [inline server/src/com/thoughtworks/go/config/update/CreateSCMConfigCommand.java]: no need for a super call, goes to super by default :)

**Pull request merged**